File: glial-local-ts/src/session_browser.ts
import type {
  BrowserSessionRecord,
  BrowserSessionRecordStore,
  BrowserSessionStorageMode,
  GripSessionStore,
  SessionSummary,
} from "./types";

export type GripSessionCatalog = GripSessionStore & BrowserSessionRecordStore;

function nowMs(): number {
  return Date.now();
}

export function createBrowserSessionId(prefix = "browser"): string {
  return `${prefix}_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 10)}`;
}

export async function ensureBrowserSessionRecord(
  store: GripSessionCatalog,
  browserSessionId: string,
  opts?: {
    title?: string;
    storageMode?: BrowserSessionStorageMode;
    glialSessionId?: string;
  },
): Promise<BrowserSessionRecord> {
  const existing = await store.getBrowserSession(browserSessionId);
  if (existing) {
    return existing;
  }
  const summary =
    opts?.glialSessionId
      ? ((await store.getSession(opts.glialSessionId)) ??
        (await store.newSession({
          session_id: opts.glialSessionId,
          title: opts.title,
        })))
      : await store.newSession({ title: opts?.title });
  const record: BrowserSessionRecord = {
    browser_session_id: browserSessionId,
    glial_session_id: summary.session_id,
    title: summary.title ?? opts?.title,
    storage_mode: opts?.storageMode ?? "local",
    last_opened_ms: nowMs(),
  };
  await store.putBrowserSession(record);
  return record;
}

export async function bindBrowserSessionToExistingSession(
  store: GripSessionCatalog,
  browserSessionId: string,
  session: SessionSummary,
  storageMode: BrowserSessionStorageMode = "local",
): Promise<BrowserSessionRecord> {
  const record: BrowserSessionRecord = {
    browser_session_id: browserSessionId,
    glial_session_id: session.session_id,
    title: session.title,
    storage_mode: storageMode,
    last_opened_ms: nowMs(),
  };
  await store.putBrowserSession(record);
  return record;
}

File: glial-local-py/src/glial_local/session_browser.py
"""Launcher-session helpers for desktop and CLI clients."""

from __future__ import annotations

from typing import Protocol
from uuid import uuid4

from .in_memory import _now_ms
from .types import (
    GripSessionStore,
    LauncherSessionKind,
    LauncherSessionRecord,
    LauncherSessionRecordStore,
    LauncherSessionStorageMode,
    NewSessionRequest,
    SessionSummary,
)


class GripSessionCatalog(GripSessionStore, LauncherSessionRecordStore, Protocol):
    """Combined session catalog used by local launcher helpers."""


def create_launcher_session_id(prefix: str = "launcher") -> str:
    return f"{prefix}_{_now_ms()}_{uuid4().hex[:8]}"


def ensure_launcher_session_record(
    store: GripSessionCatalog,
    launcher_session_id: str,
    *,
    title: str | None = None,
    storage_mode: LauncherSessionStorageMode = "local",
    session_kind: LauncherSessionKind = "local",
    glial_session_id: str | None = None,
) -> LauncherSessionRecord:
    existing = store.get_launcher_session(launcher_session_id)
    if existing is not None:
        return existing
    summary = (
        store.get_session(glial_session_id)
        if glial_session_id is not None
        else None
    ) or store.new_session(
        NewSessionRequest(session_id=glial_session_id, title=title)
    )
    record = LauncherSessionRecord(
        launcher_session_id=launcher_session_id,
        glial_session_id=summary.session_id,
        title=summary.title or title,
        storage_mode=storage_mode,
        session_kind=session_kind,
        last_opened_ms=_now_ms(),
    )
    store.put_launcher_session(record)
    return record


def bind_launcher_session_to_existing_session(
    store: GripSessionCatalog,
    launcher_session_id: str,
    session: SessionSummary,
    storage_mode: LauncherSessionStorageMode = "local",
    session_kind: LauncherSessionKind = "local",
) -> LauncherSessionRecord:
    record = LauncherSessionRecord(
        launcher_session_id=launcher_session_id,
        glial_session_id=session.session_id,
        title=session.title,
        storage_mode=storage_mode,
        session_kind=session_kind,
        last_opened_ms=_now_ms(),
    )
    store.put_launcher_session(record)
    return record
